### PR TITLE
feat: instance deletion

### LINKED ISSUE #65: Support instance deletion
## What needs to be done ?

- add feature to delete a container instance
- add feature to delete a VM 
  - Ensure it cleans up TAP interface for VM
  - Ensure it cleans up iptables rules
- Instance deletion should trigger an event sent to the scheduler saying `Destroying`, then once finished `Terminated` 

## Why is it needed ?

- We'd like to improve lifecycle of instances, unscheduling needs this feature to be done

### PR DESCRIPTION
**Pull request by AlexandreBrg**
Closes #65 

## 📑 Description

* Add nginx service into the rootfs from scripts, in order to have a reachable service on port 80
* Update controller to make a return of instance ids when creating a new instnace
* Update scheduler to make it manage the deletion of instances 
* Update riklet to make it able to delete the workloads running on their system

* Add a mermaid plugin into the mdbook documentation
* Add a page to have a reference of the workload json format

**Note:** Firepilot feature to stop a VM isn't working properly which means the `firecracker` process is only killed once the riklet process is killed. Still, `iptables` rules are deleted.

## ✅ Checks

- [ ] My pull request adheres to the code style of this project
- [ ] My code is documented
- [ ] I provided unitary tests or procedure to test my code
- [ ] My PR have a clear description of what my code is supposed to do

### CODE REVIEW

File: controller/src/api/external/routes/instance.rs [hunk 1]
```diff
@@ -106,15 +116,23 @@ pub fn delete(
...
         internal_sender
             .send(ApiChannel {
                 action: Crud::Delete,
-                workload_id: None,
-                workload_definition: None,
+                workload_id: Some(instance_def.workload_id),
+                workload_definition: Some(workload_def),
                 instance_id: Some(delete_id),
             })
             .unwrap();
```
Comment (MaloPolese) [inline controller/src/api/external/routes/instance.rs:135]: remove `unwrap()`

File: riklet/src/core.rs [hunk 2]
```diff
@@ -107,21 +111,34 @@ impl Riklet {
...
+    ///
+    /// Expected lifecycle is:
+    /// Receive delete request -> Send destroying status
+    /// -> Destroy instance & Unregister runtime -> Send terminated status
+    #[tracing::instrument(skip_all, fields(instance_id = %workload.instance_id))]
     async fn delete_workload(
         &mut self,
         workload: &InstanceScheduling,
-        runtime: DynamicRuntimeManager<'_>,
+        _runtime: DynamicRuntimeManager<'_>,
```
Comment (MaloPolese) [inline riklet/src/core.rs:123]: remove if unused

File: riklet/src/runtime/function_runtime.rs [hunk 3]
```diff
@@ -85,11 +89,12 @@ impl Runtime for FunctionRuntime {
 
         event!(Level::DEBUG, "Generate configuration for MicroVM");
         let vm_config = self.generate_microvm_config()?;
+        self.vm_config = Some(vm_config.clone());
 
         event!(Level::DEBUG, "Run microVM in a thread");
         thread::spawn(move || {
             // Thread is alive while the VM is running
-            if let Err(e) = firecracker.start(&vm_config) {
+            if let Err(e) = firecracker.start(&vm_config.clone()) {
```
Comment (MaloPolese) [inline riklet/src/runtime/function_runtime.rs:97]: useless clone

File: riklet/src/runtime/mod.rs [hunk 4]
```diff
@@ -44,7 +45,8 @@ type Result<T> = std::result::Result<T, RuntimeError>;
 
 #[async_trait]
 pub trait Runtime: Send + Sync {
-    async fn run(&mut self) -> Result<()>;
+    async fn up(&mut self) -> Result<()>;
+    async fn destroy(&self) -> Result<()>;
```
Comment (MaloPolese) [inline riklet/src/runtime/mod.rs:49]: maybe rename down ?

File: scheduler/src/state_manager/mod.rs [hunk 5]
```diff
@@ -218,6 +218,50 @@ impl StateManager {
...
+                let worker = workers.next().unwrap();
+
+                instance.set_worker(Some(worker.clone()));
+                // For now we don't check whether the instance is properly deleted, we assume it is
+                // as if we keep the destroying state, it will loop here and spam riklet of events
+                instance.set_status(ResourceStatus::Terminated);
+
+                info!("Deleting instance {}", instance.id.clone());
+
+                let _ = self
```
Comment (MaloPolese) [inline scheduler/src/state_manager/mod.rs:241]: why make an empty assignation here ?

File: scheduler/src/state_manager/mod.rs [hunk 6]
```diff
@@ -218,6 +218,50 @@ impl StateManager {
...
+                        worker.clone(),
+                        InstanceScheduling {
+                            instance_id: instance.id.clone(),
+                            action: WorkloadRequestKind::Destroy as i32,
+                            definition: serde_json::to_string(&instance.definition.clone())
+                                .unwrap(),
+                        },
+                    ))
+                    .await;
+                let _ = self
```
Comment (MaloPolese) [inline scheduler/src/state_manager/mod.rs:253]: why make an empty assignation here ?

**Review by MaloPolese — CHANGES_REQUESTED** [review]

File: controller/src/api/external/routes/instance.rs [hunk 7]
```diff
@@ -83,16 +85,24 @@ pub fn create(
...
+            &Some(instance_name),
         );
     }
 
-    Ok(tiny_http::Response::from_string("").with_status_code(tiny_http::StatusCode::from(201)))
+    Ok(
+        tiny_http::Response::from_string(serde_json::to_string(&instance_names).unwrap())
+            .with_header(tiny_http::Header::from_str("Content-Type: application/json").unwrap())
+            .with_status_code(tiny_http::StatusCode::from(201)),
+    )
```
[the remark this answers is not in the document]
Comment (AlexandreBrg) [reply controller/src/api/external/routes/instance.rs:105]: Both unwrap here are safe:

- `instance_names` is either an empty or filled vector
- `Content-Type` is safe as it is a static string

Created an issue related to this: https://github.com/rik-org/rik/issues/89

File: scheduler/src/state_manager/mod.rs (hunk 5 above)
Comment (AlexandreBrg) [reply scheduler/src/state_manager/mod.rs:241]: AFAIK it was due to memory management between Rust and async. Without it, the async task would have been dropped. Might not be relevant anymore, for safety lets' keep it.

**Review by MaloPolese — APPROVED** [review]

**Pull request merged**